python/mcap/mcap0/reader.py

```diff
@@ -0,0 +1,291 @@
+    :param end_time: if not None, messages from after this unix timestamp are not included.
+    """
+    out: List[ChunkIndex] = []
+    for chunk_index in summary.chunk_indexes:
+        if start_time is not None and chunk_index.message_end_time < start_time:
+            continue
+        if end_time is not None and chunk_index.message_start_time >= end_time:
+            continue
+        for channel_id in chunk_index.message_index_offsets.keys():
+            if topics is None or summary.channels[channel_id].topic in topics:
+                print(
+                    f"start: {chunk_index.message_start_time}, end: {chunk_index.message_end_time}"
+                )
+                out.append(chunk_index)
+                break
+    return out
+
+
+class MCAPReader(ABC):
+    @abstractmethod
+    def iter_messages(
+        self,
+        topics: Optional[Iterable[str]] = None,
+        start_time: Optional[int] = None,
+        end_time: Optional[int] = None,
+    ) -> Iterator[Tuple[Schema, Channel, Message]]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_summary(self) -> Optional[Summary]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def iter_attachments(self) -> Iterator[Attachment]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def iter_metadata(self) -> Iterator[Metadata]:
+        raise NotImplementedError()
+
+
+class SeekingReader(MCAPReader):
+    """an MCAPReader for reading out of non-seekable data sources."""
+
+    def __init__(self, stream: io.IOBase):
+        self._stream = stream
+        self._summary: Optional[Summary] = None
+
+    def iter_messages(
+        self,
+        topics: Optional[Iterable[str]] = None,
+        start_time: Optional[int] = None,
+        end_time: Optional[int] = None,
+    ) -> Iterator[Tuple[Schema, Channel, Message]]:
+        """iterates through the messages in an MCAP.
+        :param topics: if not None, only messages from these topics will be returned.
+        :param start_time: an integer nanosecond timestamp. if provided, messages logged before
+            this timestamp are not included.
+        :param end_time: an integer nanosecond timestamp. if provided, messages logged after
+            this timestamp are not included.
+        """
+        summary = self.get_summary()
+        assert summary is not None
+        if summary is None:
+            # no index available, use a non-seeking reader to read linearly through the stream.
+            return NonSeekingReader(self._stream).iter_messages(
+                topics, start_time, end_time
+            )
+        for chunk_index in _chunks_matching_topics(
+            summary, topics, start_time, end_time
+        ):
+            self._stream.seek(chunk_index.chunk_start_offset + 1 + 8, io.SEEK_SET)
+            chunk = Chunk.read(ReadDataStream(self._stream))
+            for record in breakup_chunk(chunk):
```

Comment:
So it looks like we don't support out of order messages or merging across chunks, is that right? I guess the conformance tests don't currently test that but it would be nice to eventually provide support for this across all languages.

Reply:
explained offline – TS implementation for reference https://github.com/foxglove/mcap/blob/c766292714102a58073d8420cf57023c11cb4481/typescript/core/src/v0/Mcap0IndexedReader.test.ts#L440